doobie/src/main/scala/tamer/db/package.scala

```diff
@@ -0,0 +1,137 @@
+package tamer
+
+import cats.effect.Blocker
+import com.sksamuel.avro4s.{Decoder, Encoder, SchemaFor}
+import doobie.Query0
+import doobie.hikari.HikariTransactor
+import doobie.implicits.{toDoobieStreamOps, _}
+import doobie.util.transactor.Transactor
+import eu.timepit.refined.auto._
+import fs2.Stream
+import log.effect.LogWriter
+import log.effect.zio.ZioLogWriter.log4sFromName
+import tamer.config.Config
+import tamer.db.Compat.toIterable
+import tamer.db.ConfigDb.{DbConfig, QueryConfig}
+import tamer.db.Db.{Timestamped, TimeSegment, _}
+import tamer.kafka.Kafka
+import zio.blocking.Blocking
+import zio.interop.catz._
+import zio.{Queue, Task, UIO, ZIO, _}
+
+import java.sql.SQLException
+import java.time.{Duration, Instant}
+import scala.concurrent.ExecutionContext
+
+package object db {
+  type DbTransactor  = Has[Transactor[Task]]
+  type TamerDBConfig = DbTransactor with QueryConfig
+
+  implicit final class InstantOps(private val instant: Instant) extends AnyVal {
+    def orNow: UIO[Instant] =
+      UIO(Instant.now).map {
+        case now if instant.isAfter(now) => now
+        case _                           => instant
+      }
+  }
+
+  final def mkSetupWithTimeSegment[K <: Product: Encoder: Decoder: SchemaFor, V <: Product with Timestamped: Ordering: Encoder: Decoder: SchemaFor](
+      queryBuilder: TimeSegment => Query0[V]
+  )(earliest: Instant, tumblingStep: Duration, keyExtract: V => K): Setup[K, V, TimeSegment] = {
+
+    val timeSegment = TimeSegment(earliest, earliest.plus(tumblingStep))
+
+    def stateFold(timeSegment: TimeSegment)(queryResult: QueryResult[V]): UIO[TimeSegment] =
+      if (queryResult.results.isEmpty) timeSegment.to.plus(tumblingStep).orNow.map(TimeSegment(timeSegment.from, _))
+      else {
+        val mostRecent = queryResult.results.max.timestamp
+        mostRecent.plus(tumblingStep).orNow.map(TimeSegment(mostRecent, _))
+      }
+
+    mkSetup(queryBuilder)(timeSegment, keyExtract, stateFold)
+  }
+
+  final def mkSetup[
+      K <: Product: Encoder: Decoder: SchemaFor,
+      V <: Product: Encoder: Decoder: SchemaFor,
+      S <: Product with HashableState: Encoder: Decoder: SchemaFor
+  ](
+      queryBuilder: S => Query0[V]
+  )(defaultState: S, keyExtract: V => K, stateFoldM: S => QueryResult[V] => UIO[S]): Setup[K, V, S] = {
+    val qBuilder = new QueryBuilder[V, S] {
+      override val queryId: Int               = queryBuilder(defaultState).sql.hashCode
+      override def query(state: S): Query0[V] = queryBuilder(state)
+    }
+    Setup[K, V, S](queryBuilder = qBuilder, defaultState = defaultState, keyExtract = keyExtract, stateFoldM = stateFoldM)
+  }
+
+  private[this] val logTask: Task[LogWriter[Task]] = log4sFromName.provide("tamer.db")
+
+  final def iteration[K <: Product, V <: Product, S <: Product with HashableState](
```

Comment:
is `iteration` expected to be part of the public api ? Considering we have `fetch` and `fetchWithTimeSegment` it could be `private`

Reply:
no, let's make it private for now